atc/db/job.go

```diff
@@ -934,21 +934,26 @@ func (j *job) getRunningBuildsBySerialGroup(tx Tx, serialGroups []string) ([]Bui
 }
 
 func (j *job) getNextPendingBuildBySerialGroup(tx Tx, serialGroups []string) (Build, bool, error) {
-	row := buildsQuery.Options(`DISTINCT ON (b.id)`).
+	subQuery, params, err := buildsQuery.Options(`DISTINCT ON (b.id)`).
 		Join(`jobs_serial_groups jsg ON j.id = jsg.job_id`).
 		Where(sq.Eq{
 			"jsg.serial_group":    serialGroups,
 			"b.status":            BuildStatusPending,
 			"j.paused":            false,
 			"j.inputs_determined": true,
 			"j.pipeline_id":       j.pipelineID}).
-		OrderBy("b.id ASC").
-		Limit(1).
-		RunWith(tx).
-		QueryRow()
+		ToSql()
+	if err != nil {
+		return nil, false, err
+	}
+
+	row := tx.QueryRow(`
+			SELECT * FROM (`+subQuery+`) j
+			ORDER BY COALESCE(rerun_of, id) ASC, id ASC
```

Comment:
Not sure if this would be an issue based on the[ postgres docs](https://www.postgresql.org/docs/9.0/sql-select.html#SQL-DISTINCT), which mention that `Distinct on` should be executed with a matching `order by` as the select query might have unpredictable results otherwise.

We used this to test that out and received different results for the flat and nested queries

```
CREATE TABLE test (
    id SERIAL PRIMARY KEY,
    name text NOT NULL,
    location text NOT NULL,
    baz text NOT NULL
)
```

```
insert into test (name,location, baz) values ('foo', 'nyc', 'bar');
insert into test (name,location, baz) values ('foo', 'nyc', 'bar');
insert into test (name,location, baz) values ('foo', 'tor', 'bar');
insert into test (name,location, baz) values ('foo', 'la', 'bar');
insert into test (name,location, baz) values ('moo', 'nyc', 'bar');
insert into test (name,location, baz) values ('moo', 'tor', 'bar');
insert into test (name,location, baz) values ('moo', 'la', 'bar');
insert into test (name,location, baz) values ('soo', 'la', 'bar');
```

flat query
```
SELECT DISTINCT ON (name) name, location
    FROM test
    ORDER BY name, location DESC;
```


nested query
```
select * from (SELECT DISTINCT ON (name) name, location
    FROM test) j order by name, location DESC;
```

Reply:
chatted about this, since only the build is scanned from the result, the `distinct on` behaviour shouldn't impact the build.